src/test/v1.2/advanced/channels-support-1.2.ts

```diff
@@ -0,0 +1,274 @@
+  }
+
+  createTestChannel = async (): Promise<Channel> => {
+    const channelName = `${this.testAppChannelName}.${this.getRandomId()}`;
+    return fdc3.getOrCreateChannel(channelName);
+  }
+
+  unsubscribeListeners = (): void => {
+    if (listener1 !== undefined) {
+      listener1.unsubscribe();
+      listener1 = undefined;
+    }
+
+    if (listener2 !== undefined) {
+      listener2.unsubscribe();
+      listener2 = undefined;
+    }
+  }
+
+  channelCleanUp = async (): Promise<void> => {
+    this.unsubscribeListeners();
+    await fdc3.leaveCurrentChannel();
+  }
+
+
+  closeChannelsAppWindow = async (testId: string): Promise<void> => {
+    //Tell ChannelsApp to close window
+    const appControlChannel = await broadcastAppChannelCloseWindow(testId);
+
+    //Wait for ChannelsApp to respond
+    await waitForContext("windowClosed", testId, appControlChannel);
+    await wait(constants.WindowCloseWaitTime);
+  }
+
+
+
+  initCompleteListener = async (testId: string): Promise<Context> => {
+    return waitForContext(
+      "executionComplete",
+      testId,
+      await fdc3.getOrCreateChannel("app-control")
+    );
+  }
+
+  openChannelApp = async (testId: string, channelId: string | undefined, commands: string[], historyItems: number = undefined, notify: boolean = true): Promise<void> => {
+    const channelsAppConfig: ChannelsAppConfig = {
+      fdc3ApiVersion: "1.2",
+      testId: testId,
+      channelId,
+      notifyAppAOnCompletion: notify,
+    };
+
+    if (historyItems) {
+      channelsAppConfig.historyItems = historyItems;
+    }
+
+    //Open ChannelsApp then execute commands in order
+    await fdc3.open(
+      "ChannelsApp",
+      buildChannelsAppContext(commands, channelsAppConfig)
+    );
+  }
+
+  setupAndValidateListener1 = (channel: Channel, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): void => {
+    if (channel) {
+      listener1 = channel.addContextListener(null, (context) => {
+        expect(context.type).to.be.equals(expectedContextType, errorMessage);
+        onComplete(context);
+      });
+    } else {
+      listener1 = fdc3.addContextListener(null, (context) => {
+        expect(context.type).to.be.equals(expectedContextType, errorMessage);
+        onComplete(context);
+      });
+    }
+
+    validateListenerObject(listener1);
+  }
+
+  setupAndValidateListener2 = (channel: Channel, expectedContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): void => {
+    if (channel) {
+      listener2 = channel.addContextListener(null, (context) => {
+        expect(context.type).to.be.equals(expectedContextType, errorMessage);
+        onComplete(context);
+      });
+    } else {
+      listener2 = fdc3.addContextListener(null, (context) => {
+        expect(context.type).to.be.equals(expectedContextType, errorMessage);
+        onComplete(context);
+      });
+    }
+
+    validateListenerObject(listener2);
+  }
+
```

Comment:
We could do with an alternative function here for validating that context is not received (often needed in tests):

```suggestion

validateContextIsNotReceived = (channel: Channel, listenerContextType: string, errorMessage: string, onComplete: (ctx: Context) => void): void => {
    if (channel) {
      listener1 = channel.addContextListener(listenerContextType, (context) => {
        assert.fail(errorMessage);
        onComplete(context);
      });
    } else {
      listener1 = fdc3.addContextListener(listenerContextType, (context) => {
        assert.fail(errorMessage);
        onComplete(context);
      });
    }
    validateListenerObject(listener1);
  }
  
```

Note that the error message should be customized for each usage.

Reply:
done